### ISSUE TITLE
ChildStdin does not send EOF on .shutdown()

### ISSUE THREAD

**Issue by coral** [tokio-rs/tokio#4477]
**Version**
1.16.1

```
❯ cargo tree | grep tokio
│   │   │   ├── tokio v1.16.1
│   │   │   │   └── tokio-macros v1.7.0 (proc-macro)
│   │   │   ├── tokio-util v0.6.9
│   │   │   │   └── tokio v1.16.1 (*)
│   │   │   ├── tokio v1.16.1 (*)
│   │   │   ├── tokio v1.16.1 (*)
│   │   │   └── tokio-native-tls v0.3.0
│   │   │       └── tokio v1.16.1 (*)
│   │   ├── tokio v1.16.1 (*)
│   │   ├── tokio-native-tls v0.3.0 (*)
│   ├── tokio v1.16.1 (*)
├── tokio v1.16.1 (*)
```

**Platform**

Mac OSX Monterey 12.1

Darwin 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64 x86_64

**Description**

ChildStdin does not seem to properly send EOF on `shutdown()`. I am porting some code from another project and one step is triggering file syncing using rsync as a child process, expecting a pipe of files for input.

```rust
 let mut cmd = Command::new("rsync");
cmd.args([
    "--ignore-existing",
    "-r",
    "-v",
    "--files-from=-",
    &self.source,
    &self.dest,
]);

cmd.stdout(Stdio::inherit());
cmd.stderr(Stdio::inherit());
cmd.stdin(Stdio::piped());

let mut child = cmd.spawn()?;

let mut tn = child.stdin.take();
let stdin = tn.as_mut().ok_or(Error::CouldNotGetStdin)?;

let mut filelist: String = "".to_string();
for f in files {
    filelist.push_str(&format!("{}\n", f));
}
stdin.write_all(filelist.as_bytes()).await?;

stdin.shutdown().await?;
```

What ends up happening is the process keeps waiting for the EOF without any closure. I tried by running the command manually and sending Ctrl + D (EOF) which does work. The same command also works in Go by expecting EOF

```go
// This is in Go
// EXAMPLE CODE
stdin, _ := cmd.StdinPipe()

go func() {
	defer stdin.Close()
	for _, p := range filesToSync {
		io.WriteString(stdin, p+"\n")
	}
}()
```
`stdin.Close()` in this case will send EOF upon completion whereas ChildStdin seems to not properly do this.

**Comment by Darksonn** [comment]
Thoughts @ipetkov?

**Comment by coral** [comment]
https://github.com/tokio-rs/tokio/blob/fc4deaa1d0e35d3404d1d32ddcb4b083081355c0/tokio/src/process/mod.rs#L1233

Alice Ryhl pointed out in the Discord that the trait for shutdown seems broken. From my understanding the polling of shutdown doesn't actually do anything?

**Comment by ipetkov** [comment]
The `AsyncWriteExt::shutdown` is supposed to encompass flushing any buffered state along with any other necessary graceful shutdown (e.g. like tearing down a TLS connection), but it doesn't actually close the pipe.

@coral what comes after the `shutdown().await` call? You can try explicitly dropping the `stdin` handle to make sure it gets closed, sending EOF to the child process so it can complete and exit.

```rust
let mut cmd = Command::new("rsync");
cmd.args([
    "--ignore-existing",
    "-r",
    "-v",
    "--files-from=-",
    &self.source,
    &self.dest,
]);

cmd.stdout(Stdio::inherit());
cmd.stderr(Stdio::inherit());
cmd.stdin(Stdio::piped());

let mut child = cmd.spawn()?;

let mut tn = child.stdin.take();
let stdin = tn.as_mut().ok_or(Error::CouldNotGetStdin)?;

let mut filelist: String = "".to_string();
for f in files {
    filelist.push_str(&format!("{}\n", f));
}
stdin.write_all(filelist.as_bytes()).await?;

stdin.shutdown().await?;
drop(stdin);

child.await // etc.
```

**Comment by coral** [comment]
@ipetkov 

Dropping works yes (that's how I ended up working around this problem) but I'm curious if this is the intended design of the API? As in the `stdin` basically holding onto the pipe handle for the duration of the lifetime, leaving "shutdown" basically unused? If so, is the methods for `AsyncWriter` on `ChildStdin` just assumed to basically be NOP?

**Comment by ipetkov** [comment]
@coral 

> Dropping works yes (that's how I ended up working around this problem) but I'm curious if this is the intended design of the API? 

Yep, this is the idiomatic way of handling resource cleanup in Rust, similar to closing channels, dropping the io handle leads to closing it. (in fact this is the only way to close the handle)

> As in the stdin basically holding onto the pipe handle for the duration of the lifetime, leaving "shutdown" basically unused? If so, is the methods for AsyncWriter on ChildStdin just assumed to basically be NOP?

Yeah `shutdown` and `flush` are basically NOPs on the `ChildStd{in,out,err}` handles since they don't do much on their own besides give us some type safety when working with the child process. If you were using something like a buffered reader/writer, for example, then it may be necessary to call `flush` to get all the data sent, but as the example code stands, the `shutdown` call could be elided!

**Comment by coral** [comment]
@ipetkov 

https://github.com/tokio-rs/tokio/pull/4479

Took a pass at creating a simple example for the documentation which captures a lot of what I've learned from this.

**Comment by Darksonn** [comment]
@ipetkov Are we sure this is what we want? I mean, pipes can send an EOF signal, and other types of channels (e.g. `TcpStream`, `UnixStream`) _will_ send EOF when their `poll_shutdown` method is called.

**Comment by ipetkov** [comment]
@Darksonn 

>  pipes can send an EOF signal

Is this true? AFAIU [`shutdown(3)`](https://linux.die.net/man/3/shutdown) is specific to sockets. Even the [stdlib only has shutdown for sockets](https://doc.rust-lang.org/stable/std/?search=shutdown) and not for regular pipes.

Not opposed to adding this behavior if possible, just not sure it exists

**Comment by Darksonn** [comment]
Hm, interesting. I'll have to think about what to do about that. It is a difference from std that our `AsyncWrite` trait has a method for shutdown.

**Comment by timmclean** [comment]
I think this issue should be reopened. I just spent an hour debugging a bug resulting from this behaviour. I was pretty surprised to find that `child_stdin.shutdown()` was implemented as a no-op. This doesn't seem to be documented on ChildStdin.

From the documentation for [AsyncWrite::poll_shutdown](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWrite.html#tymethod.poll_shutdown):
> Initiates or attempts to shut down this writer, returning success when the I/O connection has completely shut down.

ChildStdin's current behaviour is pretty confusing given what's documented on AsyncWrite. Based on this text, I expected `child_stdin.shutdown()` to close the underlying fd.
